### ISSUE TITLE
Integer overflow due to more than 5 simultaneous key exchanges

### ISSUE THREAD

**Issue by EinMByte** [monero-project/kovri#334]
**By submitting this issue, I confirm the following:**

- I have read and understood the [contributor guide](https://github.com/monero-project/kovri/blob/master/doc/CONTRIBUTING.md).
- I have checked that the issue I am reporting can be replicated or that the feature I am suggesting is not present.
- I have checked opened or recently closed [pull requests](https://github.com/monero-project/kovri/pulls) for existing solutions/implementations to my issue/suggestion.

*Place an X inside the bracket to confirm*
- [x] I confirm.

---

`DHKeysPairSupplier::Run()` at `transports.cc:84`:

```
 80 void DHKeysPairSupplier::Run() {
 81   LogPrint(eLogDebug, "DHKeysPairSupplier: running");
 82   while (m_IsRunning) {
 83     std::size_t num;
 84     while ((num = m_QueueSize - m_Queue.size ()) > 0)
 85       CreateDHKeysPairs(num);
 86     std::unique_lock<std::mutex> l(m_AcquiredMutex);
 87     m_Acquired.wait(l);  // wait for element gets acquired
 88   }
 89 }
```

If ` m_QueueSize < m_Queue.size()`, this causes an overflow leadings to the creation of (literally) 10 quintillion key pairs. Note that `m_QueueSize == 5` by magic.
This can happen in the follow scenario:

1. We're routing lots of  traffic, so we have at least one key exchange going on.
2. Another key exchange starts, so `m_QueueSize - m_Queue.size() == 1`. That means we'll create a new key pair. After that, we have five pairs again.
3. The new key pair gets acquired, so we now have 4 pairs again.
4. The first key exchange finishes, and the DH key pair is returned. We now have five key pairs.
5. The second key exchange finishes, and the DH key pair is returned. We now have six key pairs.
6. A new key exchange starts so we compute `m_QueueSize - m_Queue.size()`, which equals... `18446744073709551615`!

This probably went unnoticed because NTCP was pretty broken until recently (see #187), but it actually happens almost always when running with `--enable-ssu=0` (see #184).

Note also how this is a trivial DoS attack vector.

**Comment by anonimal** [comment]
Pasting from #335 for housekeeping (I should've pasted here instead):

Referencing #191 (re: NTCP, the biggest "leak" was simply the rampant generation of keypairs and not a real "leak". Fixed in d6d952c).

**Comment by anonimal** [comment]
Three notes:

1. I was able to review this and #335 in real-time in @EinMByte's fork while we collaborated over IRC
2. This issues was 100% inherited when forked
3. I added label `lol` for possibly the funniest and best bug of the year (if not the entire project)
